includes/oauth/class-mailchimp-api.php

```diff
@@ -0,0 +1,177 @@
+				'permission_callback' => function() {
+					return current_user_can( 'manage_options' );
+				},
+			]
+		);
+
+		// Verify and save Mailchimp API Key.
+		\register_rest_route(
+			NEWSPACK_API_NAMESPACE,
+			'/oauth/mailchimp',
+			[
+				'methods'             => \WP_REST_Server::EDITABLE,
+				'callback'            => [ __CLASS__, 'api_mailchimp_save_key' ],
+				'permission_callback' => function() {
+					return current_user_can( 'manage_options' );
+				},
+			]
+		);
+
+		// Delete Mailchimp API Key.
+		\register_rest_route(
+			NEWSPACK_API_NAMESPACE,
+			'/oauth/mailchimp',
+			[
+				'methods'             => \WP_REST_Server::DELETABLE,
+				'callback'            => [ __CLASS__, 'api_mailchimp_delete_key' ],
+				'permission_callback' => function() {
+					return current_user_can( 'manage_options' );
+				},
+			]
+		);
+	}
+
+	/**
+	 * Check whether we have an API key and its validity.
+	 *
+	 * @return WP_REST_Response
+	 */
+	public static function api_mailchimp_auth_status() {
+		$mailchimp_api_key = get_option( 'newspack_newsletters_mailchimp_api_key' );
+		$endpoint          = self::get_api_endpoint_from_key( $mailchimp_api_key );
+
+		if ( ! $mailchimp_api_key || ! $endpoint ) {
+			return \rest_ensure_response( [] );
+		}
+
+		$key_is_valid_response = self::is_valid_api_key( $endpoint, $mailchimp_api_key );
+
+		// delete API key if not valid.
+		if ( is_wp_error( $key_is_valid_response ) ) {
+			delete_option( 'newspack_newsletters_mailchimp_api_key' );
+		}
+
+		return $key_is_valid_response;
+	}
+
+	/**
+	 * Verify and save Mailchimp API key
+	 *
+	 * @param WP_REST_REQUEST $request call request.
+	 * @return WP_REST_Response
+	 */
+	public static function api_mailchimp_save_key( $request ) {
+		if ( isset( $request['api_key'] ) ) {
+			$endpoint = self::get_api_endpoint_from_key( $request['api_key'] );
+			if ( ! $endpoint ) {
+				return new \WP_Error( 'wrong_parameter', __( 'Invalid Mailchimp API Key.', 'newspack' ) );
+			}
+
+			$key_is_valid_response = self::is_valid_api_key( $endpoint, $request['api_key'] );
+
+			if ( ! is_wp_error( $key_is_valid_response ) ) {
+				update_option( 'newspack_newsletters_mailchimp_api_key', $request['api_key'] );
+			}
+
+			return $key_is_valid_response;
+		} else {
+			return new \WP_Error( 'missing_parameters', __( 'Missing parameters in request.', 'newspack' ) );
+		}
+	}
+
+	/**
+	 * Delete Mailchimp API Key option.
+	 *
+	 * @return WP_REST_Response
+	 */
+	public static function api_mailchimp_delete_key() {
+		delete_option( 'newspack_newsletters_mailchimp_api_key' );
+		return \rest_ensure_response( [] );
+	}
+
+	/**
+	 * Get the right datacenter from the API key and set it in the endpoint.
+	 *
+	 * @param string $api_key Mailchimp API Key.
+	 * @return string|boolean the API endpoint with the right datacenter.
+	 */
+	private static function get_api_endpoint_from_key( $api_key ) {
+		list(, $data_center) = explode( '-', $api_key );
```

Comment:
With no key saved, the `$api_key` is `false` and this line triggers a PHP notice:
```
PHP Notice:  Undefined offset: 1
```